File: data/fear_greed.py
"""Fear & Greed Index from alternative.me. Free, no key."""
from __future__ import annotations

import requests
import streamlit as st

URL = "https://api.alternative.me/fng/"


@st.cache_data(ttl=600, show_spinner=False)
def get_fear_greed(limit: int = 1) -> list[dict]:
    """Returns list with {value, value_classification, timestamp, ...}, newest first."""
    try:
        resp = requests.get(URL, params={"limit": limit, "format": "json"}, timeout=10)
        resp.raise_for_status()
        payload = resp.json()
        return payload.get("data", [])
    except Exception:
        return []


@st.cache_data(ttl=3600, show_spinner=False)
def get_fear_greed_history(days: int = 30) -> list[int]:
    """Returns list of F&G values oldest→newest."""
    try:
        resp = requests.get(URL, params={"limit": days, "format": "json"}, timeout=10)
        resp.raise_for_status()
        data = resp.json().get("data", [])
        return [int(d["value"]) for d in reversed(data)]
    except Exception:
        return []


def classify_color(value: int) -> str:
    if value < 25:
        return "#E63946"
    if value < 45:
        return "#F4A261"
    if value < 55:
        return "#E9C46A"
    if value < 75:
        return "#A8DADC"
    return "#2A9D8F"

File: lib/providers/fear-greed.ts
import "server-only";

export type FearGreed = {
  value: number;
  classification: string;
  updated: number;
};

export async function getFearGreed(): Promise<FearGreed | null> {
  try {
    const res = await fetch("https://api.alternative.me/fng/?limit=1", {
      next: { revalidate: 900 },
    });
    if (!res.ok) return null;
    const json = (await res.json()) as {
      data: { value: string; value_classification: string; timestamp: string }[];
    };
    const row = json.data?.[0];
    if (!row) return null;
    return {
      value: Number(row.value),
      classification: row.value_classification,
      updated: Number(row.timestamp) * 1000,
    };
  } catch {
    return null;
  }
}
